sso-url-helper/src/main/java/it/niedermann/nextcloud/sso/urlhelper/UrlHelper.kt

```diff
@@ -0,0 +1,132 @@
+    } catch (e: MalformedURLException) {
+        if (model.startsWith("/")) {
+            // This might be an absolute path instead of an URL, so prepend the URL of the ssoAccount, but be aware of accounts which are located in a sub directory!
+            val url = URL(ssoAccount.url.substring(0, ssoAccount.url.length - URL(ssoAccount.url).path.length) + model)
+            rewriteSpecialURLs(ssoAccount, url, fallbackPreviewSize).orElse(url)
+        } else {
+            throw IllegalArgumentException("URL must be absolute (starting with protocol and host or with a slash character).")
+        }
+    }
+}
+
+/**
+ * @return a rewritten [URL] if something special has been detected, like Share IDs, File IDs or avatars. [Optional.empty] otherwise.
+ */
+private fun rewriteSpecialURLs(ssoAccount: SingleSignOnAccount, url: URL, fallbackPreviewSize: FallbackPreviewSize): Optional<URL> {
+    // Exclude potential sub directory from url path (if Nextcloud instance is hosted at https://example.com/nextcloud)
+    val pathStartingFromNextcloudRoot = if (url.query == null) {
+        url.toString().substring(ssoAccount.url.length)
+    } else {
+        val urlStringWithoutLeadingAccount = url.toString().substring(ssoAccount.url.length)
+        urlStringWithoutLeadingAccount.substring(0, urlStringWithoutLeadingAccount.length - url.query.length - 1)
+    }
+
+    val fileId = REGEX_FILE_ID.find(pathStartingFromNextcloudRoot)?.groupValues?.get(2)
+    if (fileId != null) {
+        return if (url.query == null) {
+            Optional.of(URL("${ssoAccount.url}/index.php/core/preview?fileId=${fileId}&x=${fallbackPreviewSize.width}&y=${fallbackPreviewSize.height}&a=true"))
+        } else {
+            Optional.of(URL("${ssoAccount.url}/index.php/core/preview?fileId=${fileId}&x=${fallbackPreviewSize.width}&y=${fallbackPreviewSize.height}&a=true&${url.query}"))
+        }
+    }
+
+    val shareId = REGEX_SHARE_ID.find(pathStartingFromNextcloudRoot)?.groupValues?.get(2)
+    if (shareId != null) {
+        return if (url.query == null) {
+            Optional.of(URL("${ssoAccount.url}/index.php/s/${shareId}/download"))
+        } else {
+            Optional.of(URL("${ssoAccount.url}/index.php/s/${shareId}/download?${url.query}"))
+        }
+    }
+
+    val avatarGroupValues = REGEX_AVATAR.find(pathStartingFromNextcloudRoot)?.groupValues
+    val avatarUserId = if (avatarGroupValues?.get(2) == null) {
+        ssoAccount.name
+    } else {
+        avatarGroupValues[2]
+    }
+    val avatarSize = avatarGroupValues?.get(3)
+    if (avatarSize != null) {
+        return if (url.query == null) {
+            Optional.of(URL("${ssoAccount.url}/index.php/avatar/${avatarUserId}/${avatarSize}"))
+        } else {
+            Optional.of(URL("${ssoAccount.url}/index.php/avatar/${avatarUserId}/${avatarSize}?${url.query}"))
+        }
+    }
+
+    if (listOf("/ocs", "/index.php", "/remote.php").none { pathStartingFromNextcloudRoot.startsWith(it) }) {
```

Comment:
To exclude /ocs requests here was the only logic change I did (in a separate commit (cc44a3faa1f805b12c7ce77f448e2b7cd01cf7b9)

Reply:
That's fine, too :slightly_smiling_face: